ocsp/responder/filter_source.go

```diff
@@ -0,0 +1,142 @@
+package responder
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+type responderID struct {
+	nameHash []byte
+	keyHash  []byte
+}
+
+type filterSource struct {
+	wrapped        Source
+	hashAlgorithm  crypto.Hash
+	issuers        map[issuance.IssuerNameID]responderID
```

Comment:
I see `issuerKeyHashes` is gone. The TODO on that was:

> // TODO(#5152): Simplify this when we've fully deprecated old-style IssuerIDs.

Have we fully deprecated old-style IssuerIDs? Want to update #5152?

Also, what is the risk of that part of the change? If there's a risk we could start returning 404s for some OCSP responses, we should make it as a separate change from this big one, so it can get more scrutiny.

Reply:
We talked about this, but replying here for posterity. The previous filter worked as follows:
1) At startup time, build maps from both IssuerID and IssuerNameID to the IssuerKeyHash of each issuer cert loaded in the config
2) When the request comes in, make sure that the embedded IssuerKeyHash matches *any* of the loaded IssuerKeyHashes (by iterating over the values of one of the two maps above)
3) Look up the CertificateStatus row from the database, which comes with *either* an IssuerID or an IssuerNameID attached
4) Use that ID to look up the IssuerKeyHash in one of the two maps from above, and re-confirm that the request's IssuerKeyHash matches *that specific* IssuerKeyHash

The new approach never sees the Issuer(Name)ID from the database at all, so it does not need both maps. Instead, it:
1) At startup time, build just a single map from IssuerNameID to IssuerKeyHash from the loaded issuer certs
2) When a request comes in, make sure that the embedded IssuerKeyHash matches any of the loaded certs, and keep track of the corresponding IssuerNameID because we'll need it soon
3) Look up the CertificateStatus, throw away everything except the response itself, and parse that response so we can see the embedded ResponderName
4) Compute the IssuerNameID from the embedded ResponderName, and compare that to the IssuerNameID from step (2).

This new approach has disadvantage that it requires parsing the OCSP response, and relies on a field (ResponderName) which is guaranteed to be populated by the Go stdlib but is not guaranteed to be populated by RFC6960. But it has the advantage that it is comparing a field embedded in the request (IssuerKeyHash) to a field embedded in the response (ResponderName), rather than relying on the database's Issuer(Name)ID to be set correctly.